test/ts/unit/collateralized_contract.ts

```diff
@@ -0,0 +1,195 @@
+
+        // PAYER begins with a balance of 5 ether.
+        await mockToken.mockBalanceOfFor.sendTransactionAsync(
+            PAYER, Units.ether(5),
+        );
+
+        // TransferProxy is granted an allowance of 3 ether from PAYER.
+        await mockToken.mockAllowanceFor.sendTransactionAsync(
+            PAYER, collateralContract.address, Units.ether(5),
+        );
+
+        // Initialize ABI Decoder for deciphering log receipts
+        ABIDecoder.addABI(collateralContract.abi);
+    });
+
+    after(() => {
+        ABIDecoder.removeABI(collateralContract.abi);
+    });
+
+    describe("Initialization", () => {
+        it("points to the DebtRegistry passed in through the constructor", async () => {
+            await expect(collateralContract.debtRegistry.callAsync()).to.eventually.equal(mockRegistry.address);
+        });
+    });
+
+    describe("#collateralize", () => {
+
+      describe("invariants", () => {
+
+        const ARBITRARY_AGREEMENT_ID =
+            web3.sha3("any 32 byte hex value can represent an agreement id");
+
+        it("should throw if the amount being put up for collateral is zero", async () => {
+            await expect(collateralContract.collateralize.sendTransactionAsync(
+              ARBITRARY_AGREEMENT_ID,
+              mockToken.address,
+              new BigNumber(0), // zero collateral
+              new BigNumber(moment().add(2, 'years').unix())
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the lockup period occurs in the past", async () => {
+            await expect(collateralContract.collateralize.sendTransactionAsync(
+              ARBITRARY_AGREEMENT_ID,
+              mockToken.address,
+              new BigNumber(10),
+              new BigNumber(moment().subtract(2, 'years').unix()) // timestamp that occurs in the past.
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the agreement already has already been collateralized", async () => {
+            // We first colletaralize the contract.
+            await collateralContract.collateralize.sendTransactionAsync(
+                ARBITRARY_AGREEMENT_ID,
+                mockToken.address,
+                new BigNumber(5),
+                new BigNumber(moment().add(2, 'years').unix()),
+                { from: PAYER }
+            );
+
+            // a second attempt to collateralize the contract should fail.
+            await expect(collateralContract.collateralize.sendTransactionAsync(
+                ARBITRARY_AGREEMENT_ID,
+                mockToken.address,
+                new BigNumber(10),
+                new BigNumber(moment().add(3, 'years').unix()),
+                { from: PAYER }
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the collateral fails to transfer", async () => {
```

Comment:
For all 4 of the above invariants, I think it would be better if we changed the collateral contract to "throw" errors using the `LogError` pattern we use in `DebtKernel` -- given that Solidity doesn't ([yet](https://github.com/ethereum/solidity/issues/1686)) support throwing with a `reason` string.  This makes it easier to output human readable errors in UIs

Reply:
Also -- we should add one more invariant to the `Collateralized` contract: outputting a `LogError` if allowance to the contract is not sufficient to execute `transferFrom`